lib/Dialect/FIRRTL/LowerToRTL.cpp

```diff
@@ -133,6 +133,34 @@ static LogicalResult lower(firrtl::PadPrimOp op, ArrayRef<Value> operands,
   return success();
 }
 
+static LogicalResult lower(firrtl::CatPrimOp op, ArrayRef<Value> operands,
+                           ConversionPatternRewriter &rewriter) {
+  auto lhs = mapOperand(operands[0], op, rewriter);
+  auto rhs = mapOperand(operands[1], op, rewriter);
+  if (!lhs || !rhs)
+    return failure();
+
+  auto lhsWidth = lhs.getType().cast<IntegerType>().getWidth();
+  auto rhsWidth = rhs.getType().cast<IntegerType>().getWidth();
+
+  // TODO full argument type conversion
+  Value args [2] = {operands[0], operands[1]};
+
+  auto resultLoc = op.getLoc();
+  Type resultTypes [1] = {rewriter.getIntegerType(lhsWidth + rhsWidth)};
+  NamedAttrList attrs;
+
+  Value inserted = rewriter.create<rtl::ConcatOp>(
+    resultLoc,
+    ArrayRef<Type>(resultTypes),
+    ArrayRef<Value>(args),
```

Comment:
I think you can simplify this to:

Value inserted = rewriter.create<rtl::ConcatOp>(resultLoc, resultTypes, args);

because arrays convert to arrayref.

Reply:
With suggested code I am getting this error:
```
../llvm/mlir/include/mlir/IR/PatternMatch.h: In instantiation of ‘OpTy mlir::PatternRewriter::create(mlir::Location, Args ...) [with OpTy = circt::rtl::ConcatOp; Args = {mlir::Type*, mlir::Value*}]’:
../lib/Dialect/FIRRTL/LowerToRTL.cpp:154:35:   required from here
../llvm/mlir/include/mlir/IR/PatternMatch.h:221:16: error: no matching function for call to ‘circt::rtl::ConcatOp::build(mlir::PatternRewriter&, mlir::OperationState&, mlir::Type*&, mlir::Value*&)’
  221 |     OpTy::build(*this, state, args...);
      |     ~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~
In file included from ../include/circt/Dialect/RTL/Ops.h:18,
                 from ../lib/Dialect/FIRRTL/LowerToRTL.cpp:7:
include/circt/Dialect/RTL/RTL.h.inc:140:15: note: candidate: ‘static void circt::rtl::ConcatOp::build(mlir::OpBuilder&, mlir::OperationState&, mlir::Type, mlir::ValueRange)’
  140 |   static void build(OpBuilder &odsBuilder, OperationState &odsState, Type result, ValueRange inputs);
      |               ^~~~~
include/circt/Dialect/RTL/RTL.h.inc:140:75: note:   no known conversion for argument 3 from ‘mlir::Type*’ to ‘mlir::Type’
  140 |   static void build(OpBuilder &odsBuilder, OperationState &odsState, Type result, ValueRange inputs);
      |                                                                      ~~~~~^~~~~~
include/circt/Dialect/RTL/RTL.h.inc:141:15: note: candidate: ‘static void circt::rtl::ConcatOp::build(mlir::OpBuilder&, mlir::OperationState&, llvm::ArrayRef<mlir::Type>, mlir::ValueRange)’
  141 |   static void build(OpBuilder &odsBuilder, OperationState &odsState, ArrayRef<Type> resultTypes, ValueRange inputs);
      |               ^~~~~
include/circt/Dialect/RTL/RTL.h.inc:141:85: note:   no known conversion for argument 3 from ‘mlir::Type*’ to ‘llvm::ArrayRef<mlir::Type>’
  141 |   static void build(OpBuilder &odsBuilder, OperationState &odsState, ArrayRef<Type> resultTypes, ValueRange inputs);
      |                                                                      ~~~~~~~~~~~~~~~^~~~~~~~~~~
include/circt/Dialect/RTL/RTL.h.inc:142:15: note: candidate: ‘static void circt::rtl::ConcatOp::build(mlir::OpBuilder&, mlir::OperationState&, llvm::ArrayRef<mlir::Type>, mlir::ValueRange, llvm::ArrayRef<std::pair<mlir::Identifier, mlir::Attribute> >)’
  142 |   static void build(OpBuilder &, OperationState &odsState, ArrayRef<Type> resultTypes, ValueRange operands, ArrayRef<NamedAttribute> attributes);
      |               ^~~~~
include/circt/Dialect/RTL/RTL.h.inc:142:15: note:   candidate expects 5 arguments, 4 provided
```

Reply:
Ok, perhaps you also need to pass in the empty attribute set.

Reply:
This version worked
```
rewriter.replaceOpWithNewOp<rtl::ConcatOp>(op, resultType, ArrayRef<Value>(args));
```